File: codes/golang/15.三数之和.go
/*
 * @lc app=leetcode.cn id=15 lang=golang
 *
 * [15] 三数之和
 */

// @lc code=start
import (
	"sort"
)

func threeSum(nums []int) [][]int {
	res := [][]int{}
	counter := map[int]int{}

	for _, value := range nums {
		counter[value]++
	}
	// 去重
	uniqNums := []int{}
	for key := range counter {
		uniqNums = append(uniqNums, key)
	}

	// 排序
	sort.Ints(uniqNums)

	for i := 0; i < len(uniqNums); i++ {
		if uniqNums[i]*3 == 0 && counter[uniqNums[i]] >= 3 {
			res = append(res, []int{uniqNums[i], uniqNums[i], uniqNums[i]})
		}
		for j := i + 1; j < len(uniqNums); j++ {
			if uniqNums[i]*2+uniqNums[j] == 0 && counter[uniqNums[i]] > 1 {
				res = append(res, []int{uniqNums[i], uniqNums[i], uniqNums[j]})
			}
			if uniqNums[j]*2+uniqNums[i] == 0 && counter[uniqNums[j]] > 1 {
				res = append(res, []int{uniqNums[i], uniqNums[j], uniqNums[j]})
			}
			c := 0 - uniqNums[i] - uniqNums[j]
			if c > uniqNums[j] && counter[c] > 0 {
				res = append(res, []int{uniqNums[i], uniqNums[j], c})
			}
		}
	}
	return res
}

// @lc code=end

File: codes/python/15.三数之和.py

class Solution:
	def threeSum(self, nums: List[int]) -> List[List[int]]:
		res = []
		counter = {}
		for value in nums:
			counter[value] = counter.get(value, 0) + 1

		uniqNums = []
		for key in counter:
			uniqNums.append(key)

		uniqNums.sort()

		for i in range(len(uniqNums)):
			if uniqNums[i] * 3 == 0 and counter[uniqNums[i]] >= 3:
				res.append([uniqNums[i], uniqNums[i], uniqNums[i]])
			for j in range(i + 1, len(uniqNums)):
				if uniqNums[i] * 2 + uniqNums[j] == 0 and counter[uniqNums[i]] > 1:
					res.append([uniqNums[i], uniqNums[i], uniqNums[j]])
				if uniqNums[j] * 2 + uniqNums[i] == 0 and counter[uniqNums[j]] > 1:
					res.append([uniqNums[i], uniqNums[j], uniqNums[j]])
				c = 0 - uniqNums[i] - uniqNums[j]
				if c > uniqNums[j] and counter.get(c, 0) > 0:
					res.append([uniqNums[i], uniqNums[j], c])
		return res
